libkbfs/bare_tlf_handle.go

```diff
@@ -0,0 +1,251 @@
+// keybase1.PUBLIC_UID, then the returned handle will be for a public
+// folder. Otherwise, it will be private. PUBLIC_UID shouldn't be in
+// any list in any other case.
+func MakeBareTlfHandle(
+	writers, readers []keybase1.UID,
+	unresolvedWriters, unresolvedReaders []keybase1.SocialAssertion,
+	conflictInfo *ConflictInfo) (BareTlfHandle, error) {
+	if len(writers) == 0 {
+		return BareTlfHandle{}, ErrNoWriters
+	}
+
+	for _, w := range writers {
+		if w == keybase1.PUBLIC_UID {
+			return BareTlfHandle{}, ErrInvalidWriter
+		}
+	}
+
+	if (len(readers) + len(unresolvedReaders)) > 1 {
+		// If we have more than one reader, none of them
+		// should be the public UID.
+		for _, r := range readers {
+			if r == keybase1.PUBLIC_UID {
+				return BareTlfHandle{}, ErrInvalidReader
+			}
+		}
+	}
+
+	// TODO: Check for overlap between readers and writers, and
+	// for duplicates.
+
+	writersCopy := make([]keybase1.UID, len(writers))
+	copy(writersCopy, writers)
+	sort.Sort(uidList(writersCopy))
+
+	var readersCopy []keybase1.UID
+	if len(readers) > 0 {
+		readersCopy = make([]keybase1.UID, len(readers))
+		copy(readersCopy, readers)
+		sort.Sort(uidList(readersCopy))
+	}
+
+	var unresolvedWritersCopy []keybase1.SocialAssertion
+	if len(unresolvedWriters) > 0 {
+		unresolvedWritersCopy = make([]keybase1.SocialAssertion, len(unresolvedWriters))
+		copy(unresolvedWritersCopy, unresolvedWriters)
+		sort.Sort(socialAssertionList(unresolvedWritersCopy))
+	}
+
+	var unresolvedReadersCopy []keybase1.SocialAssertion
+	if len(unresolvedReaders) > 0 {
+		unresolvedReadersCopy = make([]keybase1.SocialAssertion, len(unresolvedReaders))
+		copy(unresolvedReadersCopy, unresolvedReaders)
+		sort.Sort(socialAssertionList(unresolvedReadersCopy))
+	}
+
+	return BareTlfHandle{
+		Writers:           writersCopy,
+		Readers:           readersCopy,
+		UnresolvedWriters: unresolvedWritersCopy,
+		UnresolvedReaders: unresolvedReadersCopy,
+		ConflictInfo:      conflictInfo,
+	}, nil
+}
+
+// IsPublic returns whether or not this BareTlfHandle represents a
+// public top-level folder.
+func (h BareTlfHandle) IsPublic() bool {
+	return len(h.Readers) == 1 && h.Readers[0].Equal(keybase1.PublicUID)
+}
+
+func (h BareTlfHandle) findUserInList(user keybase1.UID,
+	users []keybase1.UID) bool {
+	for _, u := range users {
+		if u == user {
+			return true
+		}
+	}
+	return false
+}
+
+// IsWriter returns whether or not the given user is a writer for the
+// top-level folder represented by this BareTlfHandle.
+func (h BareTlfHandle) IsWriter(user keybase1.UID) bool {
+	return h.findUserInList(user, h.Writers)
+}
+
+// IsReader returns whether or not the given user is a reader for the
+// top-level folder represented by this BareTlfHandle.
+func (h BareTlfHandle) IsReader(user keybase1.UID) bool {
+	return h.IsPublic() || h.findUserInList(user, h.Readers) || h.IsWriter(user)
+}
+
+// ResolvedUsers returns the concatenation of h.Writers and h.Readers,
```

Comment:
Looks like the caller of the existing `Users()` function, on the server side, adds an entry to the folder resolver table including the public UID.  Do you plan to have a server-side PR that replaces that behavior?  Or is it not needed / relied on?